### PR TITLE
Extend search to find more alternates

### PR DESCRIPTION
**Pull request by genadz**
fixes #2812 

# Issue

This PR implements ideas described here https://github.com/valhalla/valhalla/pull/2812 based on large set of test routes (~14K routes from `auto.txt`). So, the first thing I did was double `cost threshold` to find the upper bound of the number of alternative routes that can be found with current approach. I divided all the routes into three parts (short, medium, long) and printed some statistics for 1st and 2nd alternative routes. In more detail, I compared `cost` and `number of iterations` in the moment when alternative route was found with `cost` and `number of iterations` in the moment when first connection was found.

`cost_extend_rel` - relative cost increase needed to find alternative route after first connection was found (`sort_cost_of_alternate / sort_cost_of_first_route`).

`cost_extend_abs` - absolute cost increase needed to find alternative route after first connection was found (`sort_cost_of_alternate - sort_cost_of_first_route`).

`iter_exted_rel` - relative iterations increase needed to find alternative route after first connection was found (`num_of_iterations_when_alternate_was_found / num_of_iterations_when_first_route_was_found`).

`iter_exted_abs` - number of iterations needed to find alternative route after first connection was found (`num_of_iterations_when_alternate_was_found - num_of_iterations_when_first_route_was_found`).

#### Short routes: 10m - 1h
```
Number of all routes: 1710

percentiles	10%	20%	30%	40%	50%	60%	70%	80%	90%	95%	99%	100%

alternate 1: count = 1192 percent = 69%

cost_extend_rel	1.0075	1.0262	1.0433	1.0639	1.0831	1.1076	1.1323	1.1611	1.2065	1.2669	1.4119	1.7870
cost_extend_abs	14.0	39.0	64.3	96.0	129.5	166.0	215.7	288.8	385.9	538.0	1033.8	4816.0
iter_extend_rel	1.01	1.05	1.09	1.13	1.18	1.24	1.32	1.43	1.66	1.83	2.23	15.05
iter_extend_abs	106	335	587	863	1311	1855	2794	4083	6491	8924	14211	20127

alternate 2: count = 880 percent = 51%

cost_extend_rel	1.0337	1.0614	1.0876	1.1120	1.1309	1.1524	1.1749	1.2030	1.2453	1.2882	1.3852	1.5787
cost_extend_abs	60.9	102.0	138.0	180.0	215.0	252.0	300.0	363.8	485.1	597.1	1049.8	1407.0
iter_extend_rel	1.07	1.13	1.19	1.25	1.31	1.39	1.48	1.60	1.81	1.98	2.27	5.46
iter_extend_abs	553	1090	1548	2192	2899	3797	5003	6476	9046	11840	16292	24476
```

#### Medium routes: 1h - 5h
```
Number of all routes: 4932

percentiles	10%	20%	30%	40%	50%	60%	70%	80%	90%	95%	99%	100%

alternate 1: count = 4284 percent = 86%

cost_extend_rel	1.0057	1.0207	1.0353	1.0504	1.0668	1.0842	1.1058	1.1324	1.1732	1.2157	1.3194	1.6571
cost_extend_abs	37.0	130.0	220.0	311.0	412.0	529.8	674.1	896.4	1296.0	1701.8	2801.5	5283.0
iter_extend_rel	1.01	1.05	1.08	1.12	1.16	1.22	1.28	1.40	1.61	1.83	2.24	3.78
iter_extend_abs	304	1058	1982	3047	4308	5903	8020	11231	16207	20158	34180	152607

alternate 2: count = 3634 percent = 73%

cost_extend_rel	1.0286	1.0481	1.0652	1.0843	1.1039	1.1215	1.1438	1.1661	1.2044	1.2496	1.3604	1.5529
cost_extend_abs	175.3	301.0	411.0	521.0	636.0	756.0	907.0	1184.0	1678.8	2117.7	3328.1	6936.0
iter_extend_rel	1.06	1.11	1.15	1.20	1.25	1.33	1.41	1.53	1.73	1.97	2.33	3.33
iter_extend_abs	1551	2943	4305	5804	7583	9701	12469	15320	20789	26275	47206	157522
```

#### Long routes: > 5h
```
Number of all routes: 3745

percentiles	10%	20%	30%	40%	50%	60%	70%	80%	90%	95%	99%	100%

alternate 1: count = 3194 percent = 85%

cost_extend_rel	1.0028	1.0128	1.0226	1.0352	1.0483	1.0661	1.0849	1.1168	1.1721	1.2215	1.3945	1.7029
cost_extend_abs	67.3	337.6	627.9	962.6	1376.0	1885.2	2605.1	3833.0	7267.6	13388.3	41638.5	95045.0
iter_extend_rel	1.01	1.02	1.05	1.07	1.10	1.14	1.19	1.26	1.38	1.51	1.93	3.84
iter_extend_abs	318	1136	2225	3672	5912	9728	17901	33875	81106	140085	397335	1318635

alternate 2: count = 2566 percent = 68%

cost_extend_rel	1.0186	1.0349	1.0500	1.0652	1.0823	1.0996	1.1242	1.1539	1.2042	1.2467	1.3732	1.7277
cost_extend_abs	468.5	865.0	1279.0	1698.0	2182.5	2915.0	3948.5	5687.0	9657.0	14699.2	40308.5	83189.0
iter_extend_rel	1.04	1.08	1.11	1.16	1.20	1.25	1.31	1.39	1.52	1.65	2.00	3.31
iter_extend_abs	1681	3467	5404	8279	13119	20391	34999	62291	120744	199454	429511	992529
```

**A few things to notice:**
- `master` uses absolute cost increase that equals to `420` (~7min). It works good for short routes (finds `>90%` from all possible alternates) , worse for medium routes (finds only `50%` from all possible alternates) and bad for long routes (finds `~20%` from all possible alternates).
- we can see that small increase of `cost_exted_rel` parameter may lead to a significant increase in the alternative routes found, and the same time not to downgrade performance a lot.
- there are several cases in each table where number of iterations increased a lot. If we just increase `cost_extend_rel` parameter it may lead to a significant performance drop in worst cases. So, we should limit number of iterations somehow.

Here is the graphic that shows how cost increase influence on the number of alternative routes that will be found.
Cost_increase_to_alternates

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by genadz** [comment]
I tried new cost threshold to extend search that equals to `1.1 * cost_of_first_route` and used iterations threshold delta equal to `100000`.
On `auto.txt` routes total time increased by `~20%` (from `4m56s` to `5m55s`). There is timings for short, medium and long routes:

#### short
```
master: alternates: 1st = 64%, 2nd = 45%
branch: alternates: 1st = 65%, 2nd = 47%

percentiles	5%	10%	20%	50%	90%	95%	99%	100%
master
response_time	0.008s	0.012s	0.019s	0.040s	0.105s	0.170s	0.363s	1.659s
branch
response_time	0.009s	0.012s	0.019s	0.040s	0.103s	0.172s	0.350s	1.600s
```
#### medium
```
master: alternates: 1st = 49%, 2nd = 24%
branch: alternates: 1st = 65%, 2nd = 41%

percentiles	5%	10%	20%	50%	90%	95%	99%	100%
master
response_time	0.024s	0.034s	0.047s	0.077s	0.217s	0.303s	0.496s	0.847s
branch
response_time	0.025s	0.035s	0.050s	0.084s	0.249s	0.345s	0.553s	1.158s
```
#### long
```
master: alternates: 1st = 20%, 2nd = 4%
branch: alternates: 1st = 65%, 2nd = 39%

percentiles	5%	10%	20%	50%	90%	95%	99%	100%
mater
response_time	0.045s	0.057s	0.079s	0.177s	0.674s	1.038s	1.665s	2.609s
branch
response_time	0.053s	0.071s	0.105s	0.227s	0.919s	1.238s	1.894s	2.931s
```

Timings for short routes stayed the same. For medium and long routes response time increased by `~10-30%` on `p50, p90, p99`

**Comment by kevinkreiser** [comment]
@genadz that matches pretty much exactly what i expected. let me know when you are ready for us to look at it!

**Comment by genadz** [comment]
> @genadz that matches pretty much exactly what i expected. let me know when you are ready for us to look at it!

yeah, of course! I just wanted to try some other parameters (like `1.15 * cost_of_first_route`) and compare with current results.

**Comment by genadz** [comment]
as for me, value `cost_ext_rel=0.1` looks reasonable. We got significant improvements in number of alternates and didn't downgrade performance too much.

**Comment by dgearhart** [comment]
@genadz  The original threshold was added so that we could find the fastest route or fastest alternative route for routes like this example: 
```
-j '{"locations":[{"lat":40.242828,"lon":-76.707028,"street":"Stoverdale Road"},{"lat":40.534634,"lon":-78.402096,"street":"East 25th Avenue"}],"costing":"auto","units":"miles"}'
```
Does it find a route or alternate route that takes I 76 West and I 99 North?
/Cc @kevinkreiser @gknisely

**Comment by genadz** [comment]
> @genadz The original threshold was added so that we could find the fastest route or fastest alternative route for routes like this example:
> 
> ```
> -j '{"locations":[{"lat":40.242828,"lon":-76.707028,"street":"Stoverdale Road"},{"lat":40.534634,"lon":-78.402096,"street":"East 25th Avenue"}],"costing":"auto","units":"miles"}'
> ```
> 
> Does it find a route or alternate route that takes I 76 West and I 99 North?
> /Cc @kevinkreiser @gknisely

why do you think that my changes break this request ? actually, yes, I see the route in alternatives that has I 76 West and I 99 North

**Comment by dgearhart** [comment]
> why do you think that my changes break this request ? actually, yes, I see the route in alternatives that has I 76 West and I 99 North
> 
> I use relative cost threshold , but not less than `420`

Wanted to makes sure it was the first or second route found with recent updates - it has not been working properly recently.

**Comment by genadz** [comment]
I conducted one more experiment: it's easy to notice that if we found all alternates it's not necessary to continue and we can stop bidirectional search.
I compared number of iterations that were spent to find **shortest route, 1st alternative, 2nd alternative** and **actual** number of iterations with `cost_extend_rel=1.1` and `cost_extend_rel=1.15`.

For `cost_extend_rel=1.1` if we stopped just after we found **1st alternative** we would save `6.8%` of iterations, and if we stopped just after we found **2nd alternative** we would save `3.6%` of iterations.

For `cost_extend_rel=1.15` if we stopped just after we found **1st alternative** we would save `9%` of iterations, and if we stopped just after we found **2nd alternative** we would save `5.5%` of iterations.

@kevinkreiser @dgearhart

**Comment by genadz** [comment]
@kevinkreiser it's ready for review)

**Review by PureW — COMMENTED** [review]
Nice, the data looks nice and the code seems ok to me. I'd be interested in what @kevinkreiser and @dnesbitt61 thinks

File: src/thor/bidirectional_astar.cc [hunk 1]
```diff
@@ -551,6 +562,10 @@ BidirectionalAStar::GetBestPath(valhalla::Location& origin,
   travel_type_ = costing_->travel_type();
   access_mode_ = costing_->access_mode();
 
+  desired_paths_count_ = 1;
+  if (options.has_alternates() && options.alternates())
+    desired_paths_count_ += options.alternates();
```
Comment (mandeepsandhu) [inline src/thor/bidirectional_astar.cc:567]: Do we need to protect against bad values here or is `options.alternates()` sanitized upstream of this call?

**Comment by mandeepsandhu** [comment]
Change look good to me. Is there value in adding unit tests to mock generating alternate routes (with low values set for cost & iteration threshold since gurka map will be quite small) and seeing the effect of this change?

File: src/thor/bidirectional_astar.cc (hunk 1 above)
Comment (kevinkreiser) [reply src/thor/bidirectional_astar.cc:567]: its sanitized in loki

File: src/thor/bidirectional_astar.cc [hunk 2]
```diff
@@ -753,12 +773,20 @@ bool BidirectionalAStar::SetForwardConnection(GraphReader& graphreader, const BD
...
+  // Set thresholds to extend search
+  if (cost_threshold_ == std::numeric_limits<float>::max()) {
+    float sortcost = std::max(pred.sortcost() + cost_diff_, opp_pred.sortcost());
+    if (desired_paths_count_ == 1) {
+      cost_threshold_ = sortcost + kThresholdDelta;
+    } else {
+      cost_threshold_ = sortcost + std::max(kAlternativeCostExtend * sortcost, kThresholdDelta);
+      iterations_threshold_ =
+          edgelabels_forward_.size() + edgelabels_reverse_.size() + kAlternativeIterationsDelta;
+    }
```
Comment (kevinkreiser) [inline src/thor/bidirectional_astar.cc:789]: seems like the coverage check is wrong, there are definitely unit tests that would have `desired_paths_count > 1`

Comment (kevinkreiser) [reply src/thor/bidirectional_astar.cc:789]: oh i see just randomly the other direction gets hit by the test instead of this one nevermind

**Review by kevinkreiser — DISMISSED** [review]

**Comment by kevinkreiser** [comment]
This looks good to me, thanks for taking the idea and expanding upon it!

**Comment by kevinkreiser** [comment]
>  I think, this can be done in a separate PR, because it requires significant refactoring of BidirectionalAstar class.

I agree thats a good idea. The refactoring will have to be calling `FormPath` iteratively as we find and rule out connections so that we can stop early after the desired paths have been met. It makes total sense to me. I've always wanted to rewrite it that way any way so thank you very much for tackling it! Let me know if you need any help or need to brainstorm anything

**Comment by genadz** [comment]
> > I think, this can be done in a separate PR, because it requires significant refactoring of BidirectionalAstar class.
> 
> I agree thats a good idea. The refactoring will have to be calling `FormPath` iteratively as we find and rule out connections so that we can stop early after the desired paths have been met. It makes total sense to me. I've always wanted to rewrite it that way any way so thank you very much for tackling it! Let me know if you need any help or need to brainstorm anything

Sure, I will ping you in some PR or issue about that. (Thanks for the review!)

**Comment by genadz** [comment]
> Change look good to me. Is there value in adding unit tests to mock generating alternate routes (with low values set for cost & iteration threshold since gurka map will be quite small) and seeing the effect of this change?

We define constants in `.cpp` file, so it will be difficult to reach `iterations_thershold` in gurka tests. What we can do with gurka is to build long route with alternatives differing by more than 7min.

**Review by mandeepsandhu — APPROVED** [review]
LGTM. Thanks for adding the tests

**Pull request merged**